sys/dev/altera/jtag_uart/altera_jtag_uart_cons.c

```diff
@@ -108,48 +116,61 @@ SYSCTL_UINT(_hw_altera_jtag_uart, OID_AUTO, ac_poll_delay,
  * low-level console is used for polled read while the TTY driver is also
  * looking for input.  Probably we should also share buffers between layers.
  */
-#define	MIPS_XKPHYS_UNCACHED_BASE	0x9000000000000000
-
-typedef	uint64_t	paddr_t;
 
-static inline vaddr_t
-mips_phys_to_uncached(paddr_t phys)            
+/*
+ * Get memory-mapped JTAG UART register address in
+ * uncached physically mapped memory.
+ */
+static inline char *
+mips_aju_register_addr(vm_paddr_t regoff)
 {
 
-	return (phys | MIPS_XKPHYS_UNCACHED_BASE);
+#ifdef __CHERI_PURE_CAPABILITY__
+	return (cheri_setoffset(aju_io_capability, regoff));
+#else
+	return (MIPS_PHYS_TO_XKPHYS_UNCACHED(BERI_UART_BASE + regoff));
+#endif
```

Comment:
Maybe `aju_io_capability` should be an `aju_io_pointer` that gets set to BERI_UART_BASE for !purecap, and then we just use normal pointer arithmetic?

Reply:
good point